core/interfaces/branch.py

```diff
@@ -524,43 +529,46 @@ def get_selected_branch_name(self, selection, line):
         branch_name = segments[1]
         local_region = self.view.get_regions("git_savvy_interface.branch_list")[0]
         if local_region.contains(selection):
-            return branch_name
+            return (None, branch_name)
 
         remotes = self.get_remotes()
         for remote_name in remotes:
             remote_region = self.view.get_regions("git_savvy_interface.branch_list_" + remote_name)
             if remote_region and remote_region[0].contains(selection):
-                return "{}/{}".format(remote_name, branch_name)
+                return (remote_name, branch_name)
+
+    def create_branches_strs(self, branches):
+        branches_strings = set()
+        for branch in branches:
+            if branch[0] is None:
+                branches_strings.add(branch[1])
+            else:
+                branches_strings.add("{}/{}".format(branch[0], branch[1]))
+        return branches_strings
 
 
-class GsBranchesFetchAndMergeCommand(TextCommand, GitCommand):
+class GsBranchesFetchAndMergeCommand(GsBranchesMergeSelectedCommand):
 
     """
     Fetch from remote and merge fetched branch into active branch.
     """
 
-    def run(self, edit):
-        sublime.set_timeout_async(self.run_async, 0)
-
     def run_async(self):
         self.interface = ui.get_interface(self.view.id())
-        selection, line = self.interface.get_selection_line()
-        if not line:
-            return
 
-        segments = line.strip("▸ ").split(" ")
-        branch_name = segments[1]
+        branches = self.get_selected_branches()
+        for branch in branches:
+            if branch[0] is None:
+                # is remote is not set it is a local branch and can't be fetched
+                continue
+            sublime.status_message("Fetching from `{}`...".format(branch[0]))
+            self.fetch(remote=branch[0])
```

Comment:
We're fetching all branches from remote here correct? So this may fetch multiple times from same branch without this being necessary (slowing down the operation). Try instead something like:

```python
remotes_to_fetch = set(filter(None, (b[0] for b in branches)))
for remote in remotes_to_fetch:
    sublime.status_message("Fetching from `{}`...".format(remote))
    self.fetch(remote=remote)
    sublime.status_message("Fetch from {} complete.".format(remote))
```

And the row `sublime.status_message("Fetch and merge complete.")` can be moved after the call to `self.merge`.